File: scripts/optimize_images.py
from pathlib import Path

from PIL import Image


ROOT = Path(__file__).resolve().parents[1]
SOURCE_DIR = ROOT / "assets" / "images"
OUTPUT_DIR = SOURCE_DIR / "optimized"
MAX_WIDTH = 1400
QUALITY = 78
TARGETS = {
    "blog-life.png",
    "blog-marathon-simple.png",
    "project-people-counter.jpg",
    "icon-highlight-ai.png",
    "icon-highlight-community.png",
    "icon-highlight-data.png",
    "icon-highlight-ml.png",
    "NLP_MediMapAI.png",
    "pneumonia-cnn-research.jpg",
}


def optimize_image(path: Path) -> None:
    with Image.open(path) as image:
        image = image.convert("RGB")
        image.thumbnail((MAX_WIDTH, MAX_WIDTH), Image.Resampling.LANCZOS)
        output_path = OUTPUT_DIR / f"{path.stem}.webp"
        image.save(output_path, "WEBP", quality=QUALITY, method=6)


def main() -> None:
    OUTPUT_DIR.mkdir(exist_ok=True)
    optimized = 0

    for filename in sorted(TARGETS):
      path = SOURCE_DIR / filename
      if path.exists() and path.stat().st_size > 0:
          optimize_image(path)
          optimized += 1

    print(f"Optimized {optimized} referenced images.")


if __name__ == "__main__":
    main()

File: scripts/optimize-images.mjs
import { mkdir, readdir, stat } from "node:fs/promises";
import path from "node:path";
import { fileURLToPath } from "node:url";
import sharp from "sharp";

const root = path.dirname(path.dirname(fileURLToPath(import.meta.url)));
const sourceDir = path.join(root, "assets/images");
const outputDir = path.join(root, "assets/images/optimized");
const supported = new Set([".jpg", ".jpeg", ".png"]);
const targets = new Set([
  "blog-life.png",
  "blog-marathon-simple.png",
  "project-people-counter.jpg",
  "icon-highlight-ai.png",
  "icon-highlight-community.png",
  "icon-highlight-data.png",
  "icon-highlight-ml.png",
  "NLP_MediMapAI.png",
  "pneumonia-cnn-research.jpg"
]);

await mkdir(outputDir, { recursive: true });

const entries = await readdir(sourceDir, { withFileTypes: true });
let optimizedCount = 0;

for (const entry of entries) {
  if (!entry.isFile()) continue;
  const ext = path.extname(entry.name).toLowerCase();
  if (!supported.has(ext)) continue;
  if (!targets.has(entry.name)) continue;

  const sourcePath = path.join(sourceDir, entry.name);
  const sourceStat = await stat(sourcePath);
  if (sourceStat.size === 0) continue;

  const outputName = `${path.basename(entry.name, ext)}.webp`;
  const outputPath = path.join(outputDir, outputName);

  await sharp(sourcePath)
    .rotate()
    .resize({ width: 1400, height: 1400, fit: "inside", withoutEnlargement: true })
    .webp({ quality: 78, effort: 5 })
    .toFile(outputPath);

  optimizedCount += 1;
}

console.log(`Optimized ${optimizedCount} images to ${path.relative(root, outputDir)}.`);
